app/session.js

```diff
@@ -21,8 +21,60 @@ try {
 
 const envFromConfig = config.getConfig().env || {};
 
+// Max duration to batch session data before sending it to the renderer process.
+const BATCH_DURATION_MS = 16;
+
+// Max size of a session data batch. Note that this value can be exceeded by ~4k
+// (chunk sizes seem to be 4k at the most)
+const BATCH_MAX_SIZE = 200 * 1024;
```

Comment:
I tried different values but this one (200k) had the best throughput. Using, e.g. 100k, can lower the throughput by ~25% :fire:

Reply:
A future improvement could determine this value dynamically

Reply:
Few notes from my side:
Keep in mind that the 16ms stack on top of the input roundtrip latency and ppl might start complaining about lousy terminal response while typing. In the xterm.js demo app I found 5 ms helping alot while not yet being noticeable, 15 ms already introduced a tiny lag when typing really fast.

4k is the default pty buffer size on many linux distros, but you cannot rely on that (can be changed for every pty and system wide). BSDs (and macOS) use an adaptive buffer size typically ranging from 16k to 64k depending on the current data pressure.

Reply:
Btw 200k seems quite big to me, you might want to test this with `yes`, it triggers a very exp. 'lineFeed' event in the terminal for every second byte. Additionally the event can be hooked by external handler, that might introduce even more workload (not sure if you allow plugins to consume it).

Reply:
> Keep in mind that the 16ms stack on top of the input roundtrip latency and ppl might start complaining about lousy terminal response while typing. In the xterm.js demo app I found 5 ms helping alot while not yet being noticeable, 15 ms already introduced a tiny lag when typing really fast.

I had the same concern (main reason I use kitty) but I haven't felt any lag so far. We'll have to keep an eye on feedback.

> 4k is the default pty buffer size on many linux distros, but you cannot rely on that (can be changed for every pty and system wide). BSDs (and macOS) use an adaptive buffer size typically ranging from 16k to 64k depending on the current data pressure.

Good point, I actually have to remove the "Note..." part, It was happening in my origin implementation, but it didn't feel right to go over a "max", so I fixed it to never exceed `BATCH_MAX_SIZE`

Reply:
> what strategies would you use for dynamic computation of this value? Maybe we can create an RFC issue?

I was thinking the parser could generate "back pressure". Basically "Hey, the chunks you are sending are too big for me to parse and render in one frame, slow down". This could be less important if we [run the parser on a WebWorker](https://github.com/xtermjs/xterm.js/issues/1515)

Reply:
> Btw 200k seems quite big to me, you might want to test this with yes, it triggers a very exp. 'lineFeed' event in the terminal for every second byte

I just tested `yes` and the buffering definitely doesn't help (drops FPS from ~50 to ~30). Basically 200k is a good value for `find ~` which is arguably a more common output than `yes`. This is all more evidence for implementing @jerch's proposal here: https://github.com/xtermjs/xterm.js/pull/1818#issuecomment-445457198

Reply:
Yeah I think its better suited in `Terminal.write`. We already do xon/xoff flow control there, imho its better to keep it in one place. Otherwise every buffering step involved would have to implement flow control on its own to get a working back pressure chain, which gets quite cumbersome with node's async nature.